[CPU] add Float8OpaqueTensor for dynamic float8 act float8 weight

Xia-Weiwen opened the pull request:
**Summary**
We split the original big PR https://github.com/pytorch/ao/pull/2505 into the following smaller ones:
- https://github.com/pytorch/ao/pull/3039 (relanded by https://github.com/pytorch/ao/pull/3059)
- https://github.com/pytorch/ao/pull/3052
- And this PR https://github.com/pytorch/ao/pull/3075, which as the Float8OpaqueTensor for dynamic float8 act float8 weight quantization on CPU

**Test plan**
```
pytest -sv test/quantization/quantize_/workflows/float8/test_float8_opaque_tensor.py
```

Xia-Weiwen commented:
CC @mingfeima for review. Thanks.

Xia-Weiwen commented:
Hi @mingfeima @jerryzh168 @andrewor14 Could you please review this PR? Thanks.

mingfeima commented on test/quantization/quantize_/workflows/float8/test_float8_opaque_tensor.py:
```diff
@@ -0,0 +1,164 @@
+    @common_utils.parametrize("x_dim", [2, 3])
+    @common_utils.parametrize("bias", [True, False])
+    @common_utils.parametrize("bs", [1, 160])
+    @common_utils.parametrize(
+        "x_granularity",
+        [PerTensor(), PerRow(), PerGroup(32), PerGroup(64), PerGroup(128)],
+    )
```
does torch.ao support per block quantization, e.g. deepseek style?

Xia-Weiwen replied:
Thanks for the comment. The supported granularity varies among different quantization methods in Torchao. For float8 da8w8 on CPU, it does not support the block-wise quantization used in DeepSeek.

mingfeima commented on torchao/quantization/quantize_/workflows/float8/float8_opaque_tensor.py:
```diff
@@ -0,0 +1,227 @@
+
+class Float8OpaqueTensor(TorchAOBaseTensor):
+    """
+    Float8 dynamic activation float8 weight on CPU. The weight tensor is reordered to a blocked layout
```
should it be `Float8 dynamic quantized float8 weight on CPU`

Xia-Weiwen replied:
The expression here is interpreted as "Float8 dynamic activation" "float8 weight" on CPU, which means activation is dynamic quantized to float8 and weight is static quantized to float8. This is aligned with other parts in Torchao.

mingfeima commented on torchao/quantization/quantize_/workflows/float8/float8_opaque_tensor.py:
```diff
@@ -0,0 +1,227 @@
+    [block_k, block_n] may be further reordered to VNNI layout depending on supported CPU ISA.
+
+    Tensor Attributes:
+        qdata: Reordered float8 weight on CPU with shape = [N/block_n, K/block_k, block_k, block_n].
```
are we computing with float32 here, as the weight is not packed in vnni2 format.

Xia-Weiwen replied:
We are computing with bf16 or fp8, depending on ISA. The exposed shape does not have the VNNI dimension but the memory layout is VNNI-2 or VNNI-4 if ISA is supported.

jerryzh168 commented on torchao/quantization/quant_api.py:
```diff
@@ -1807,17 +1837,20 @@ def _float8_dynamic_activation_float8_weight_quantize_tensor(weight, config):
-    _check_hardware_support(granularity)
+    is_cpu = weight.device.type == "cpu"
+    if not is_cpu:
+        _check_hardware_support(granularity)
```
can you move this to version 1? and then version 2 can do this check in the tensor itself probably

jerryzh168 commented on torchao/quantization/quant_api.py:
```diff
@@ -1807,17 +1837,20 @@ def _float8_dynamic_activation_float8_weight_quantize_tensor(weight, config):
 
-    if not _fp8_mm_compat(weight):
+    if not is_cpu and not _fp8_mm_compat(weight):
         # TODO(future PR): this should really throw an exception instead of silently
         # not doing what the user asked
         return weight
 
-    if isinstance(weight_granularity, PerRow):
+    if not is_cpu and isinstance(weight_granularity, PerRow):
         assert weight.dtype == torch.bfloat16, (
             "PerRow quantization only works for bfloat16 precision input weight"
         )
```
also these checks, I feel we can move these to version 1 branch for now and deprecate later, we can move the checks to tensors

jerryzh168 commented on torchao/float8/inference.py:
```diff
@@ -204,28 +205,41 @@ def _normalize_granularity(
         ]
     ],
+    supported_granularities: tuple[FP8Granularity] = (PerTensor, PerRow),
+    support_different_granularities: bool = False,
```
I think just create `_normalize_and_validate_granularities` for both float8_tensor.py and float8_opaque_tensor.py in their own files without sharing anything seems to be the cleanest

jerryzh168 replied:
oh actually looks like https://github.com/pytorch/ao/blob/838dcebb13be7cd7b7ae5d1d8c4e674c89a5d7b8/torchao/quantization/qat/fake_quantize_config.py#L422 needs a pure normalize function

I think you can do the following:

torchao/quantization/quantize_/workflows/float8/utils.py
_normalize_granularity (make a single element to be a tuple)

torchao/quantization/quantize_/workflows/float8/float8_tensor.py
_validate_granluarity (per row, per tensor)

torchao/quantization/quantize_/workflows/float8/float8_opaque_tensor.py
_validate_granluarity (per row, per tensor, per block)

Xia-Weiwen commented:
Hi @jerryzh168 It's been awhile since last update of this PR because of reversions of prior PRs. I have rebased this PR. And for the `_normalize_granularities` thing we discussed above (https://github.com/pytorch/ao/pull/3075#discussion_r2433348592), I have added a method `_normalize_and_check_granularities` to `Float8OpaqueTensor`. This brings minimal changes and keeps the code clean. Could you please review this PR again? Thanks.

jerryzh168 commented on test/quantization/quantize_/workflows/float8/test_float8_opaque_tensor.py:
```diff
@@ -0,0 +1,164 @@
+            )
+
+
+common_utils.instantiate_parametrized_tests(TestFloat8OpaqueTensor)
```
nit: seems that we can add this  to `TestFloat8OpaqueTensor` class as a decorator

jerryzh168 commented on torchao/quantization/quant_api.py:
```diff
@@ -1834,12 +1862,12 @@ def _float8_dynamic_activation_float8_weight_quantize_tensor(weight, config):
         return weight
 
-    if isinstance(weight_granularity, PerRow):
+    if not is_cpu and isinstance(weight_granularity, PerRow):
```
same here

jerryzh168 commented on torchao/quantization/quant_api.py:
```diff
@@ -1849,6 +1877,7 @@ def _float8_dynamic_activation_float8_weight_quantize_tensor(weight, config):
             "Config Deprecation: version 1 of Float8DynamicActivationFloat8WeightConfig is deprecated and will no longer be supported in a future release, please use version 2, see https://github.com/pytorch/ao/issues/2649 for more details"
         )
 
+        _check_hardware_support(granularity)
```
this function name seems too general, but we can improve this later

Xia-Weiwen commented on test/quantization/quantize_/workflows/float8/test_float8_opaque_tensor.py:
```diff
@@ -0,0 +1,164 @@
+    @common_utils.parametrize("x_dim", [2, 3])
+    @common_utils.parametrize("bias", [True, False])
+    @common_utils.parametrize("bs", [4, 128])
+    def test_dynamic_float8_linear_ref(self, dtype, x_dim, bias, bs):
```
It tests the fallback path in the kernel. I have updated the name and added a comment to explain this. Thanks.

Xia-Weiwen commented on test/quantization/quantize_/workflows/float8/test_float8_opaque_tensor.py:
```diff
@@ -0,0 +1,164 @@
+
+from torchao import quantize_
+from torchao.quantization import PerGroup, PerRow, PerTensor
+from torchao.quantization.quant_api import (
```
Updated. Thanks.

Xia-Weiwen commented on torchao/quantization/quant_api.py:
```diff
@@ -1834,12 +1862,12 @@ def _float8_dynamic_activation_float8_weight_quantize_tensor(weight, config):
             return weight
 
-    elif not _fp8_mm_compat(weight):
+    elif not is_cpu and not _fp8_mm_compat(weight):
```
Updated. Thanks.

Xia-Weiwen commented on torchao/quantization/quant_api.py:
```diff
@@ -1898,12 +1939,6 @@ def _float8_dynamic_activation_float8_weight_transform(
     *,
     parameter_name: str = "weight",
 ):
-    assert is_sm_at_least_89() or is_MI300(), (
```
It's probably due to rebase conflicts. I have added this back. Thanks.

Xia-Weiwen commented on torchao/quantization/quant_api.py:
```diff
@@ -1898,12 +1939,6 @@ def _float8_dynamic_activation_float8_weight_transform(
-    assert is_sm_at_least_89() or is_MI300(), (
-        "Float8 dynamic activation quantization is only supported on CUDA>=8.9 and MI300+"
-    )
-    if config.set_inductor_config:
```
It's probably due to rebase conflicts. I have added this back. Thanks.

Xia-Weiwen commented on torchao/quantization/quantize_/workflows/float8/float8_opaque_tensor.py:
```diff
@@ -0,0 +1,273 @@
+            float8_dtype=torch.float8_e4m3fn,
+            block_size=act_block_size,
+        )
+        act_mat = _quantize_affine_float8(act_mat, act_scale, torch.float8_e4m3fn)
```
Thanks for the pointer. However, `_choose_quant_func_and_quantize_tensor` does the following:
```python
    if isinstance(quant_kwargs, QuantizeTensorToFloat8Kwargs):
        return Float8Tensor.from_hp(...)
```
Unfortunately, `Float8OpaqueTensor` also uses `QuantizeTensorToFloat8Kwargs` so it cannot distinguish them.
Besides, in the implementation of `Float8Tensor`, activation is quantized by `Float8Tensor.from_hp` to a `Float8Tensor` and then unwrapped to get the quantized tensor data for computation. And this part of logic is not exposed to users. So, I feel that it's unnecessary to use `Float8OpaqueTensor.from_hp` to quantize the activation then unwrap it. It looks good to quantize it with `_quantize_affine_float8`.
What do you think? If you want `Float8OpaqueTensor` to be aligned with `Float8Tensor`, we may need to define a counterpart of `QuantizeTensorToFloat8Kwargs` for `Float8OpaqueTensor` so that we can distinguish them. Thanks.

jerryzh168 replied:
yeah it's optional, if you feel it's better to not use it, it's OK as well.

main thing is using this will reduce duplication and make it easier to adapt in the future

Xia-Weiwen replied:
Thanks for the suggestion. Since input tensors should be plain, we can reuse `QuantizeTensorToFloat8Kwargs` and `_choose_quant_func_and_quantize_tensor` here. I have updated this part.

Xia-Weiwen commented on torchao/quantization/quantize_/workflows/float8/float8_opaque_tensor.py:
```diff
@@ -0,0 +1,273 @@
+        packed_weight,
+        scale,
+        bias.float() if bias is not None else bias,  # requires bias to be float
+        torch.float,  # out_dtype
```
Updated. Thanks.

jerryzh168 commented on torchao/quantization/quant_api.py:
```diff
@@ -1716,6 +1718,22 @@ def _input_activation_quant_func_fp8(
     return activation
 
 
+def _input_activation_quant_cpu_fp8(
```
is this function used?

Xia-Weiwen replied:
Thanks. Removed.

Xia-Weiwen commented on torchao/testing/model_architectures.py:
```diff
@@ -64,12 +64,12 @@ def __init__(
 
     # Note: Tiny-GEMM kernel only uses BF16 inputs
-    def example_inputs(self, batch_size=1):
+    def example_inputs(self, batch_size=1, dtype=None):
```
I have removed this. Thanks.

Xia-Weiwen commented on test/quantization/quantize_/workflows/float8/test_float8_opaque_tensor.py:
```diff
@@ -0,0 +1,150 @@
+                return
+        device = "cpu"
+        m = ToyTwoLinearModel(256, 256, 256, dtype, device, bias).eval()
+        example_inputs = m.example_inputs(batch_size=bs, dtype=dtype)
```
Updated. Thanks.

jerryzh168 commented on torchao/quantization/quantize_/workflows/float8/float8_opaque_tensor.py:
```diff
@@ -0,0 +1,273 @@
+        f"Shapes of input and weight do not match, input:{input_tensor.shape}, weight: {weight_tensor.shape}"
+    )
+
+    act_mat = input_tensor.contiguous()
```
OK sounds good
